include/unifex/any_object.hpp

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/scope_guard.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/detail/any_heap_allocated_storage.hpp>
+#include <unifex/detail/type_erasure_builtins.hpp>
+#include <unifex/detail/vtable.hpp>
+#include <unifex/detail/with_forwarding_tag_invoke.hpp>
+#include <unifex/detail/with_type_erased_tag_invoke.hpp>
+
+#include <cstdint>
+#include <memory>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex
+{
+  template <
+      std::size_t InlineSize,
+      std::size_t InlineAlignment,
+      bool RequireNoexceptMove,
+      typename DefaultAllocator,
+      typename... CPOs>
+  struct _any_object {
+    class type;
+  };
+
+  template <
+      std::size_t InlineSize,
+      std::size_t InlineAlignment,
+      bool RequireNoexceptMove,
+      typename DefaultAllocator,
+      typename... CPOs>
+  class _any_object<
+      InlineSize,
+      InlineAlignment,
+      RequireNoexceptMove,
+      DefaultAllocator,
+      CPOs...>::type : private with_type_erased_tag_invoke<type, CPOs>... {
+    // Pad size/alignment out to allow storage of at least a pointer.
+    static constexpr std::size_t padded_alignment =
+        InlineAlignment < alignof(void*) ? alignof(void*) : InlineAlignment;
+    static constexpr std::size_t padded_size =
+        InlineSize < sizeof(void*) ? sizeof(void*) : InlineSize;
```

Comment:
I originally did use `std::max` here which is constexpr since C++14 but was concerned about the overhead of pulling in the entire `<algorithm>` header just to be able to use it. Maybe we can put our own constexpr `max` in the `unifex` namespace somewhere?

I also went through various iterations of trying to pad the padded_size up to the next multiple of `padded_alignment` but wasn't sure whether or not that was worthwhile or not. It might take up extra padding space that would otherwise be used by classes that inherit from this one.

Reply:
A `unifex::max` seems like a good balance between expressing intent and minimizing include overhead.

Reply:
Any thoughts on where such a utility should live in libunifex?

Reply:
We have `<unifex/std_concepts.hpp>`; perhaps there should be a `<unifex/std_algorithms.hpp>`?